### DISCUSSION TITLE
Receiving Works; Can't seem to Send (Hardware Issue?)

### DISCUSSION THREAD

**Discussion by raisinbread** [bengtmartensson/Infrared4Arduino#78] [Q&A]
Could use a little help here. I'm using a TSOP4856 receiver, and a TSAL6100 IR emitter/LED. 

https://www.vishay.com/docs/82489/tsop322.pdf
https://www.vishay.com/docs/81009/tsal6100.pdf

Using the [IrReceiverSampler_SenderPwm](https://github.com/bengtmartensson/Infrared4Arduino/blob/master/examples/IrReceiverSampler_SenderPwm/IrReceiverSampler_SenderPwm.ino) example, I can record signals that are transmitted from remote controls in my home. So I feel like that part of the setup works. 

However, I can't seem to send a signal (that is in turned recorded as received). I'm driving the LED using a transistor so I can use a bit more current from an Arduino Uno 5V power output pin (rather than the limited current on the digital output pin). When I trigger a send in my code, I can use the camera on my phone to verify that the LED is firing - and it is. I've also changed the `send()` call to use a different frequency (`56000U`) to make sure it works with my receiver.

I also see the "Sending a signal from PROGMEM!" messages in the serial console.

I'm a little stumped - I can see the LED in a device remote from my camera, and I can see the LED in my circuit blinking in my camera, but my code never registers the data from my sender circuit. Any tips or things I should track down first?

**Comment by bengtmartensson** [comment]
You are trying to send a untested signal using untested hardware. Try do separate the problems. I would recommend trying to send a simple, well-known signal (for example a Philips or Sony TV) first, to verify the hardware. Possibly you (temporarily?) take out the transistor -- after all, 40mA makes a more powerful sender than most standard remotes.

Please post a diagram or a photo of the sender.

Please also post the signal you learned.

Note that the used example is mainly intended for testing by developers; it is not otherwise not very useful.

**Reply by raisinbread** [reply]
@bengtmartensson Thanks for taking a look. I have tried it with and without the transistor, but still no luck. In either configuration, I can see the LED light up (using my phone camera). I'll attach pictures of both configurations.

![IMG_5329](https://user-images.githubusercontent.com/42132/129385784-7ee32549-8336-489d-acc9-bba44ad073fa.jpg)
![IMG_5330](https://user-images.githubusercontent.com/42132/129385804-19a38355-206e-4b30-8121-681ba919f4b8.jpg)

The signal I learned is from my BenQ Projector. It just happened to be the first one I picked up. The signal output (per `receiver->dump()`) looks like this:

```
+9050 -4400 +550 -600 +500 -650 +500 -600 +500 -650 +500 -600 +500 -650 +500 -600 +550 -600 +500 -600 +450 -700 +300 -850 +200 -3200 +100 -2150 +450 -650 +500 -600 +500 -1750 +500 -650 +500 -1750 +500 -650 +500 -1750 +500 -600 +500 -650 +500 -600 +500 -650 +450 -1800 +450 -700 +500 -1750 +500 -600 +500 -1750 +500 -1750 +500 -1750 +500 -30100
```

Definitely noted that the example is just for testing. I'm mostly just trying to get a working setup so I can build from there. Thanks so much for taking a look.

**Comment by bengtmartensson** [comment]
Use a resistor in series with the LED.

Without it, you will basically have a short-circuit when the LED is on, having "interesting" consequences for the reliable operation of the CPU.

The learned signal is really bad too.

**Reply by raisinbread** [reply]
@bengtmartensson Oh right - I've done that on and off too. Sometimes I'd read that for short bursts, it's fine not to, but adding it doesn't seem to make a difference.

![IMG_5332](https://user-images.githubusercontent.com/42132/129407600-ad3e526b-afa8-49b0-9fa2-9b147259ef3f.JPG)

I'd actually be super happy even if I got a bad or sloppy signal, but right now the board doesn't register anything at all.

Here's also the code I'm using, in case there's a problem here as well. It's modified from the example, partly for simplicity (I realize it may not be portable, but that's fine by me):

```
// This sketch combines sending and receiving.

#include <Arduino.h>
#include <IrReceiverSampler.h>
#include <IrSenderPwm.h>

static constexpr pin_t RECEIVE_PIN = 8U;
static constexpr size_t BUFFERSIZE = 200U;
static constexpr uint32_t BAUD = 115200UL;

// NEC(1) 122 29 with no repetition; powers on many Yamaha receivers
static const microseconds_t array[]
#if HAS_FLASH_READ
        PROGMEM
#endif
        = {
    9024, 4512, 564, 564, 564, 1692, 564, 564, 564, 1692, 564, 1692,
    564, 1692, 564, 1692, 564, 564, 564, 1692, 564, 564, 564, 1692,
    564, 564, 564, 564, 564, 564, 564, 564, 564, 1692, 564, 1692, 564,
    564, 564, 1692, 564, 1692, 564, 1692, 564, 564, 564, 564, 564, 564,
    564, 564, 564, 1692, 564, 564, 564, 564, 564, 564, 564, 1692, 564,
    1692, 564, 1692, 564, 39756
};

static const IrSequence* irSequence =
#if HAS_FLASH_READ
        IrSequence::readFlash(array, sizeof (array) / sizeof (microseconds_t));
#else
        new IrSequence(array, sizeof (array) / sizeof (microseconds_t));
#endif
static IrReceiver *receiver;
static IrSender *sender;

void setup() {
    Serial.begin(BAUD);
    while(!Serial);
    receiver = IrReceiverSampler::newIrReceiverSampler(BUFFERSIZE, RECEIVE_PIN);
    sender = IrSenderPwm::getInstance(true);
    Serial.println(F("Listening!"));
    receiver->enable();
}

void loop() {

    if (Serial.read() != -1) {
      #if HAS_FLASH_READ
        Serial.println(F("Sending a signal from PROGMEM!"));
      #else
        Serial.println(F("Sending a signal!"));
      #endif
      sender->send(*irSequence, 56000U);
    } else {
      receiver->receive(); // combines enable, loop, disable
      if (receiver->isEmpty()) {
        Serial.println(F("No data."));
      } else {
        receiver->dump(Serial);
        Serial.println();
      }
    }
    
}
```
I _do_ see the LED light up under this configuration, after I send something to the console (although there a bit of a delay?), but I don't see any data dumped like I do when I press a button on my projector remote.

**Comment by bengtmartensson** [comment]
If i see correctly, you are using a 36 ohm resistor (orange-blue-black). That gives a current of some (5-2.0)/35 = 0.086 A, chich is more than double of the allowed 40mA. You may have damaged your board already!

**Reply by raisinbread** [reply]
@bengtmartensson I don't quite understand. The desired current for the LED is closer to 100mA - peak is up to 200ma (and I'm at 0.086 A). And if the current is too high, which part(s) of the board are getting damaged? 

https://www.vishay.com/docs/81009/tsal6100.pdf

**Comment by yesyesuk** [comment]
@raisinbread The Arduino board. Each pin can only handle up to 40mA. There is also a limit for all pins together, but I forgot what that was.

You should use a transistor to drive the IR LED.

**Reply by raisinbread** [reply]
@yesyesuk - that's what I understand... and it's how I originally had it configured, but I think the aim was to simplify things to understand what might be going wrong.

**Reply by bengtmartensson** [reply]
You can send IR with or without a driver transistor, as long as you keep within the specifications. 40mA can make a quite powerful remote, see e.g. [this](http://www.harctoolbox.org/arduino_nano.html). What you _cannot do_ is to effectively connect a LED between +5V and ground without a resistor. [Here](http://www.harctoolbox.org/arduino_nano_part2.html#Appendix.+Computing+the+current+through+the+IR+LEDs) is some stuff I wrote on how to compute the resistor.

It is also a good idea to put a small capacitator (100nF or so, not electrolytic) between + and ground on the receiver. [Example](https://github.com/Arduino-IRremote/Arduino-IRremote/issues/168#issuecomment-898528390)

**Comment by raisinbread** [comment]
Thanks for the advice, I will definitely give it a try, but it still confuses me why the receiver doesn't receive any signal at all. I appreciate your time and advice.

**Comment by raisinbread** [comment]
@bengtmartensson In your view, the code is okay then? I should be able to receive and send at the same time?

**Comment by raisinbread** [comment]
@bengtmartensson I made some progress! I looks like `receive()` has an embedded disable, so I rewrote `loop()` to look more like this:

```
void loop() {

    receiver->enable(); // combines enable, loop, disable

    while(!receiver->isReady()) {
      if (Serial.read() != -1) {
        #if HAS_FLASH_READ
          Serial.println(F("Sending a signal from PROGMEM!"));
        #else
          Serial.println(F("Sending a signal!"));
        #endif
        sender->send(*irSequence, 56000U);
      }
    }

    receiver->disable();
    
    if (receiver->isEmpty()) {
      Serial.println(F("No data."));
    } else {
      receiver->dump(Serial);
      Serial.println();
    }
    
}
```

Now I get some data when I press a key on the serial terminal:

```
Sending a signal from PROGMEM!
+65535 -30100
```

So I think I'm actually getting data, but the signal doesn't match what I'm trying to send (and that first timing looks like the maximum value). Curious if you have any ideas.

**Comment by bengtmartensson** [comment]
I strongly suggest that you solve *one problem at a time*. Start by getting the example IrReceiverSampler to work. 

Note that there is also a small probability that you have damaged the board by "sending" with no/too small  resistor.

     Sending a signal from PROGMEM!
    +65535 -30100

This is noise.
